net/server.go

```diff
@@ -427,16 +427,6 @@ func verifyRequest(header *pb.Header, body proto.Marshaler) (pid peer.ID, err er
 		err = status.Error(codes.InvalidArgument, "bad request")
 		return
 	}
-	payload, err := body.Marshal()
-	if err != nil {
-		err = status.Error(codes.Internal, err.Error())
-		return
-	}
-	ok, err := header.PubKey.Verify(payload, header.Signature)
```

Comment:
Sweet! I wonder if we can pull out the peer ID of the caller from the gRPC connection. The underlying transport is libp2p. That way, the caller doesn't have to send their public key at all. The issue is that without the signature we don't have a way to verify they are giving us a public key they own.

Reply:
Sounds like a good idea! Without signature the whole `verifyRequest()` call is meaningless it seems. I guess it's time for me to step back and have a grasp of the whole picture.

Reply:
Yep, without the signature, we're just using `verifyRequest()` to get the caller's `peer.ID`. Here's some more context:

The peer API in threads is used to exchange thread logs and records between peers. It's a gRPC API, but the underlying protocol is libp2p. We set that up here: https://github.com/textileio/go-threads/blob/master/net/net.go#L161 (the libp2p host is passed to a handy `gostream` package.

Basically, we want to know the `peer.ID` of the caller in all the peer API server methods.

When a peer dials another peer, they setup the gRPC dial context to use their host: https://github.com/textileio/go-threads/blob/master/net/server.go#L48. That `getLibp2pDialer()` method is what leverages the _caller's_ libp2p host. 

The inner closure of `getLibp2pDialer()` has access to the host (and thus the host's `peer.ID`: `s.net.host.ID()`), but I'm not sure how that info can be extracted on the receiver's end. It might be worth inspecting the gRPC context in the server-side methods (where `verifyRequest()` is currently being called) to see if `peer.ID` gets added there. If not, we could consider PRing `gostream`.

Reply:
It seems that other than logging, the peer ID is used only for `CallQueue.Schedule`, specifically [`ffQueue.Schedule`](https://github.com/textileio/go-threads/blob/master/net/queue/ff.go#L150), as the index to the set of peers and inflight calls (together with thread ID). If we can somehow find another way to uniquely identify peers within the process, we should be able to get rid of the public key. Or if having bogus peer ID doesn't cause security concerns we can just use it for now, or even exchange peer IDs directly rather than public keys.

Without having a deep understanding of the code base, I have a hunch that not having assumption of the underlying protocol (libp2p) may be useful to, for example, test the logic on mocked `net.Conn`s, but I could be very naive.

Reply:
Agreed! Spoofing the `peer.ID` (which is currently derived from the public key of the caller) _could_ be an attack vector if the caller wanted to mess up the receivers batching strategy. I looked at the incoming context to see what info is getting added there. Looks like the incoming context has a `peerKey` (gRPC peer, not libp2p peer), which via `gostream` _is actually_ the libp2p `peer.ID`. Handy! So, we can just replace calls to `verifyRequest` with something like this:

```
import (
	...
	grpcpeer "google.golang.org/grpc/peer"
)

...

func peerIDFromCtx(ctx context.Context) (peer.ID, error) {
	ctxPeer, ok := grpcpeer.FromContext(ctx)
	if !ok {
		return "", errors.New("unable to identify stream peer")
	}
	pid, err := peer.Decode(ctxPeer.Addr.String())
	if err != nil {
		return "", fmt.Errorf("parsing stream peer id: %v", err)
	}
	return pid, nil
}
```

Reply:
Wow thanks for digging into that for me! I should've inspected the context in the first place. Now that `Header` will have nothing, guess I can remove it entirely in the protobuf definition.

Reply:
No problem! This felt like an additional thing beyond the initial task... so didn't want to bog you down!